sklearn/cross_validation.py

```diff
@@ -332,6 +332,112 @@ def __len__(self):
         return self.n_folds
 
 
+def disjoint_label_folds(y, n_folds=3):
+    """Creates folds where a same label is not in two different folds
+    
+    Parameters
+    ----------
+    y: iterable of shape (n_samples, )
+        contains an id for each sample
+        The folds are built so that the same id doesn't appear in two different folds
+    
+    n_folds: int, default is 3
+        number of folds to split the data into
+        
+    Returns
+    -------
+    folds: numpy array of shape (n_samples, )
+        array of integers between 0 and (n_folds - 1)
+        folds[i] contains the folds to which sample i is assigned.
+        
+    Notes
+    -----
+    The folds are built by distributing the labels by frequency of appearance.
+    """
+    labels = np.array(y)
+    unique_labels = np.unique(labels)
+    
+    # number of occurrence of each label (its "weight")
+    weight_per_label = sorted([(sum(labels == label), label) for label in unique_labels])
```

Comment:
(your solution currently has quadratic complexity in the length of ``unique_labels``).